File: epi_judge_python/is_number_palindromic.py
import math

from test_framework import generic_test


def is_palindrome_number(x: int) -> bool:
    if x <= 0:
        return x == 0

    num_digits = math.floor(math.log10(x)) + 1
    most_significant_digit_mask = 10 ** (num_digits - 1)

    for i in range(num_digits // 2):
        if x // most_significant_digit_mask != x % 10:
            return False

        x %= most_significant_digit_mask
        x //= 10
        most_significant_digit_mask //= 100

    return True


if __name__ == '__main__':
    exit(
        generic_test.generic_test_main('is_number_palindromic.py',
                                       'is_number_palindromic.tsv',
                                       is_palindrome_number))

File: epi_judge_cpp/is_number_palindromic.cc
#include "test_framework/generic_test.h"

bool IsPalindromeNumber(int x) {
    if (x < 0) return false;

    const int digitsCount = static_cast<int>(floor(log10(x))) + 1;
    int mask = static_cast<int>(pow(10, digitsCount - 1));

    for (int i = 0; i < digitsCount / 2; ++i) {
        if (x / mask != x % 10) return false;

        x %= mask;
        x /= 10;
        mask /= 100;
    }

    return true;
}

int main(int argc, char *argv[]) {
    std::vector<std::string> args{argv + 1, argv + argc};
    std::vector<std::string> param_names{"x"};
    return GenericTestMain(args, "is_number_palindromic.cc",
                           "is_number_palindromic.tsv", &IsPalindromeNumber,
                           DefaultComparator{}, param_names);
}
